Source/Core/Runtime/Machines/Machine.cs

```diff
@@ -495,6 +503,57 @@ protected void Pop()
             this.IsPopInvoked = true;
         }
 
+        /// <summary>
+        /// Starts a timer that sends a <see cref="TimerElapsedEvent"/> to this machine after the specified
+        /// due time. The timer accepts an optional payload to be used during timeout. The timer can be stopped
+        /// by invoking the dispose method (not thread-safe and should only happen from within a handler of the
+        /// machine that owns the timer). The machine automatically disposes the timer after it timeouts.
+        /// </summary>
+        /// <param name="dueTime">The amount of time to wait before sending the first timeout event.</param>
+        /// <param name="name">Optional unique name of the timer.</param>
+        /// <param name="payload">Optional payload of the timeout event.</param>
+        /// <returns>The machine timer.</returns>
+        protected IMachineTimer StartTimer(TimeSpan dueTime, string name = null, object payload = null)
+        {
+            // The specified due time and period must be valid.
+            this.Assert(dueTime.TotalMilliseconds >= 0, $"Machine '{this.Id}' registered a timer with a negative due time.");
+            var timer = base.Runtime.CreateMachineTimer(this, dueTime, Timeout.InfiniteTimeSpan, name, payload);
+
+            lock (this.Timers)
+            {
+                this.Timers.Add(timer);
+            }
+
+            return timer;
+        }
+
+        /// <summary>
+        /// Starts a periodic timer that sends a <see cref="TimerElapsedEvent"/> to this machine
+        /// after the specified due time, and then repeats after each specified period. The timer
+        /// accepts an optional payload to be used during timeout. The timer can be stopped by
+        /// invoking the dispose method (not thread-safe and should only happen from within a
+        /// handler of the machine that owns the timer).
+        /// </summary>
+        /// <param name="dueTime">The amount of time to wait before sending the first timeout event.</param>
+        /// <param name="period">The time interval between timeout events.</param>
+        /// <param name="name">Optional unique name of the timer.</param>
+        /// <param name="payload">Optional payload of the timeout event.</param>
+        /// <returns>The machine timer.</returns>
+        protected IMachineTimer StartPeriodicTimer(TimeSpan dueTime, TimeSpan period, string name = null, object payload = null)
+        {
+            // The specified due time and period must be valid.
+            this.Assert(dueTime.TotalMilliseconds >= 0, $"Machine '{this.Id}' registered a periodic timer with a negative due time.");
+            this.Assert(period.TotalMilliseconds >= 0, $"Machine '{this.Id}' registered a periodic timer with a negative period.");
+            var timer = base.Runtime.CreateMachineTimer(this, dueTime, period, name, payload);
+
+            lock (this.Timers)
```

Comment:
Why is a lock needed here? Machine methods can only be access by the machine. I don't see why there would be concurrent accesses to `Timers`

Reply:
Agreed. This is now removed. Also modified the APIs so that timer disposal can only happen via a `StopTimer` machine API, the user cannot directly `Dispose` the timer now, which can be potentially error prone.